### ISSUE TITLE
Running replays directly from external web source

### ISSUE THREAD

**Issue by netnazgul** [OpenRA/OpenRA#13810]
> [18:34] <+pchote> the uri handlers already distinguish that
> [18:34] <+pchote> from the playtest on
> [18:35] <+pchote> for games they are e.g. openra-ra-playtest-20170722://127.0.0.1:1234
> [18:35] <+pchote> and for replays they could be openra-replay-ra-playtest-20170722://my.website/path/to.orarep

**Comment by jrb0001** [comment]
The last suggestion obviously doesn't work because the HTTP, HTTPS or maybe another supported protocols has to be specified.

**Comment by pchote** [comment]
`openra-replay-ra-playtest-20170722://` *is* the protocol.

**Comment by jrb0001** [comment]
and which protocol will be used to download the data then?

**Comment by netnazgul** [comment]
Valid point is here:

> [20:13] jaZz_KCS: if you would put for example "openra-replay-ra-playtest-20170722://64h.mine.nu/mIRROR/SOME_REPLAY_FILE.orarep" into your browser or click on such a link., your browser would dl the replay file and afterwards load it in ORA with the respective version. No streaming, the file has to be DL completely before load
> [20:14] jrb0001: nope, the browser would simply start an external application with the URL as an argument.
> [20:14] jrb0001: (at least the common ones)

**Comment by pchote** [comment]
Yes, that's the idea.  OpenRA would query the remote resource using a `WebClient` and stream the data as it is downloaded into the playback buffer similarly to how spectating a live game works.  The protocols supported by the in-game fetching is left to whoever implements this - the simple option would be to force http, but there's no reason why we couldn't try https first and then fall back to http on failure.

**Comment by jrb0001** [comment]
Forcing HTTP would be way worse than forcing HTTPS because of security reasons. The proper solution would allow specifying the protocol in the scheme part of the URI.

**Comment by pchote** [comment]
Sure, that works too. `openra-replay-ra-playtest-20170722://` is already a mouthful, so `openra-replay-secure-ra-playtest-20170722://` isn't that much worse.

**Comment by jrb0001** [comment]
A pattern like `openra-replay-ra-playtest-20170722+http://` and `openra-replay-ra-playtest-20170722+https://` would be better in my opinion because it can be extended to support additional protocols (`openra-replay-ra-playtest-20170722+ftp://` and others) easily.

**Comment by Mailaender** [comment]
See https://github.com/OpenRA/OpenRA/pull/12909 which removed the local ones on how to register new MIME types.

**Comment by pchote** [comment]
#13138, #13133, and https://github.com/OpenRA/OpenRALauncherOSX/commit/9a403b82ab2835fcd689f0c4534aa299040dcc28 provide better examples on how to register new per-mod-version handlers.
